ci/authn-gce/get_gce_id_tokens.sh

```diff
@@ -0,0 +1,144 @@
+    wait
+  fi
+}
+
+check_if_gce_instance_is_running() {
+  if [ "${INSTANCE_EXISTS}" = "1" ]; then
+    local status=$(gcloud compute instances describe "${INSTANCE_NAME}" --zone=$GCE_ZONE --format="value(status)")
+
+    if [ "${status}" = "RUNNING" ]; then
+      echo "-- Instance '${INSTANCE_NAME}' is running..."
+      INSTANCE_RUNNING=1
+    else
+      error_exit "-- Instance '${INSTANCE_NAME}' is NOT running!"
+    fi
+  else
+    error_exit "-- WARNING! Google Compute Engine instance: '${INSTANCE_NAME}' not found!"
+  fi
+}
+
+check_if_gce_instance_exists() {
+  gcloud compute instances describe ${INSTANCE_NAME} --zone="${GCE_ZONE}" --format="value(name)"
+  local error_level="$?"
+
+  if [ "${error_level}" = "0" ]; then
+    echo "-- GCE instance '${INSTANCE_NAME}' found."
+    INSTANCE_EXISTS=1
+  else
+    echo "${PROGNAME}: ${LINENO} GCE instance '${INSTANCE_NAME}' NOT found in zone: '$GCE_ZONE'!"
+
+    # check if there are any instances in the given zone
+    instances=$(gcloud compute instances list --limit=1 --format="value(name)" --zones=$GCE_ZONE)
+
+    if [ ${#instances} -eq 0 ]; then
+      error_exit "No instances found in zone: '$GCE_ZONE'"
+    else
+      gcloud compute instances list --zones=$GCE_ZONE \
+      --format="table[box,title='Instances in zone: $GCE_ZONE'](name,machine_type.basename(),status)"
+    fi
+    exit 1
+  fi
+}
+
+set_zone() {
+  if [ -z "$1" ]; then
+    echo "-- Using default zone '${GCE_ZONE}'"
+  else
+    GCE_ZONE=$(sed 's/.*=//' <<< $1)
+    echo "-- Using zone '${GCE_ZONE}'"
+  fi
+}
+
+check_instance_name_arg() {
+  if [ -z "$1" ]; then
+    echo "${PROGNAME}: ${LINENO} GCE instance name is required. Usage: ./${PROGNAME} [GCE_INSTANCE_NAME]"
+    gcloud compute instances list --zones="${GCE_ZONE}" \
+    --format="table[box,title='Instances in zone: $GCE_ZONE'](name,machine_type.basename(),status)"
+    exit 1
+  else
+    INSTANCE_NAME=$1
+    echo "-- GCE instance name is set to: '${INSTANCE_NAME}'."
+  fi
+}
+
+ensure_gcloud_is_installed() {
+  COMMAND=gcloud
+
+  if ! command -v $COMMAND &>/dev/null; then
+    error_exit "-- ${COMMAND} could not be found."
+  fi
+  echo "-- ${COMMAND} command exists"
+}
+
+echo "-- ------------------------------------------ --"
+echo "-- Generate Google Cloud GCE Identity tokens --"
+echo "-- ------------------------------------------ --"
+
+echo "-- Verifying 'gcloud' is installed..."
+ensure_gcloud_is_installed
+echo "-- Verifying CGE instance name..."
+check_instance_name_arg "$1"
+echo "-- Setting CGE zone..."
+set_zone "$2"
+echo "-- Checking if GCE instance: '${INSTANCE_NAME}' exists in zone: '$GCE_ZONE'..."
+check_if_gce_instance_exists
+echo "-- Checking if GCE instance: '${INSTANCE_NAME}' in status running..."
+check_if_gce_instance_is_running
+echo "-- Generate tokens and writing to files under '${TOKENS_OUT_DIR_PATH}'..."
+get_tokens_into_files
+echo "-- Done."
```

Comment:
as this is the main part of the script, i think it is best that it goes to the top. you can put it in a `main` function up top and then call it here. this way when the reader sees this file they can immediately read the main logic.

Reply:
also - please order the function implementations by the order they are run. it makes it easier to scroll through the script

Reply:
They are from bottom to top

Reply:
i can see that and i think that it is less readable.